File: 7.Go版每日一题/297. 二叉树的序列化与反序列化.go
package leetcode297

import (
	"strconv"
	"strings"
)

func main() {}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var SEP = ','
var NULL = '#'

type Codec struct{}

func Constructor() (_ Codec) {
	return
}

func (this *Codec) serialize(root *TreeNode) string {
	sb := &strings.Builder{}
	dfs(root, sb)

	return sb.String()
}
func dfs(root *TreeNode, sb *strings.Builder) {
	if root == nil {
		sb.WriteString("#,")
		return
	}
	sb.WriteString(strconv.Itoa(root.Val))
	sb.WriteByte(',')
	dfs(root.Left, sb)
	dfs(root.Right, sb)
}

func (Codec) deserialize(data string) *TreeNode {
	sp := strings.Split(data, ",")
	var build func() *TreeNode
	build = func() *TreeNode {
		if sp[0] == "#" {
			sp = sp[1:]
			return nil
		}
		val, _ := strconv.Atoi(sp[0])
		sp = sp[1:]
		var res = new(TreeNode) //&TreeNode{}
		res.Val = val
		res.Left = build()
		res.Right = build()
		return res // return &TreeNode{val, build(), build()}
	}
	return build()
}

File: 7.Go版每日一题/297. 二叉树的序列化与反序列化.ts
class TreeNode {
	val: number
	left: TreeNode | null
	right: TreeNode | null
	constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
		this.val = val === undefined ? 0 : val
		this.left = left === undefined ? null : left
		this.right = right === undefined ? null : right
	}
}
let SEP = ','
let NULL = '#'

function serialize(root: TreeNode | null): string {
	let sb = []
	serializeHelper(root, sb)
	return sb.toString()
}

function serializeHelper(root: TreeNode | null, sb: Array<number | string>) {
	if (root == null) {
		sb.push(NULL)
		return
	}
	sb.push(root.val) //前序遍历位置
	serializeHelper(root.left, sb)
	serializeHelper(root.right, sb)
}

//反序列化
function deserialize(data: string): TreeNode | null {
	// 将字符串转化成列表
	let nodes = []
	for (let s of data.split(SEP)) {
		nodes.push(s)
	}
	return deserializeHelper(nodes)
}

function deserializeHelper(nodes: Array<number | string | null>) {
	if (nodes.length === 0) return null

	let first = nodes.shift() //前序遍历位置
	if (first === NULL) return null
	let root = new TreeNode(parseInt(first as string, 10))
	root.left = deserializeHelper(nodes)
	root.right = deserializeHelper(nodes)
	return root
}
